staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go

```diff
@@ -225,10 +255,72 @@ func (r *proxyHandler) updateAPIService(apiService *apiregistrationv1api.APIServ
 		serviceNamespace: apiService.Spec.Service.Namespace,
 		servicePort:      *apiService.Spec.Service.Port,
 		serviceAvailable: apiregistrationv1apihelper.IsAPIServiceConditionTrue(apiService, apiregistrationv1api.Available),
+		defaultSNI:       transportConfig.TLS.ServerName,
+	}
+
+	r.updateHandlingInfo(&newInfo)
+}
+
+func (r *proxyHandler) updateHandlingInfo(handlingInfo *proxyHandlingInfo) {
+	handlingInfo.proxyRoundTripper, handlingInfo.transportBuildingError = transport.New(handlingInfo.transportConfig)
+	if handlingInfo.transportBuildingError != nil {
+		klog.Warning(handlingInfo.transportBuildingError.Error())
+	}
+	r.handlingInfo.Store(*handlingInfo)
+}
+
+func newFallbackRoundTripper(proxyHandler *proxyHandler, roundTripper http.RoundTripper) *fallbackRoundTripper {
+	return &fallbackRoundTripper{
+		proxyHandler:         proxyHandler,
+		originalRoundTripper: roundTripper,
+	}
+}
+
+var _ utilnet.RoundTripperWrapper = &fallbackRoundTripper{}
+
+type fallbackRoundTripper struct {
+	proxyHandler         *proxyHandler
+	originalRoundTripper http.RoundTripper
+}
+
+func (f *fallbackRoundTripper) WrappedRoundTripper() http.RoundTripper {
+	return f.originalRoundTripper
+}
+
+func (f *fallbackRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	value := f.proxyHandler.handlingInfo.Load()
```

Comment:
Because of this later load, the handlingInfo used to produce the proxyRoundTripper isn't necessarily the same one used retrieved here.  That race seems unintentional and not-good.  I expected to see a matched set of round trippers.

Reply:
i attempted to address this in https://github.com/kubernetes/kubernetes/pull/123808/commits/81af319c9a542a69e6f172aa504586ee4479b844